src/main/scala/mesosphere/marathon/core/instance/Instance.scala

```diff
@@ -45,58 +49,79 @@ case class Instance(
 
     // TODO(PODS): make sure state transitions are allowed. maybe implement a simple state machine?
     op match {
-      case InstanceUpdateOperation.ForceExpunge(_) =>
-        InstanceUpdateEffect.Expunge(this)
-
       case InstanceUpdateOperation.MesosUpdate(instance, status, mesosStatus, now) =>
-        // TODO(PODS): calculate the overall state afterwards
         val taskId = Task.Id(mesosStatus.getTaskId)
-        val effect = tasks.find(_.taskId == taskId).map { task =>
-          task.update(TaskUpdateOperation.MesosUpdate(status, mesosStatus))
-        }.getOrElse(TaskUpdateEffect.Failure(s"$taskId not found in $instanceId"))
-
-        effect match {
-          case TaskUpdateEffect.Update(newTaskState) =>
-            val updated: Instance = updatedInstance(newTaskState, now)
-            InstanceUpdateEffect.Update(updated, Some(this))
-
-          case TaskUpdateEffect.Expunge(newTaskState) =>
-            val updated: Instance = updatedInstance(newTaskState, now)
-            // TODO(PODS-BLOCKER): remove TaskUpdateEffect.Expunge and calculate the InstanceUpdateEffect based on
-            // the instance status
-            InstanceUpdateEffect.Expunge(updated)
-
-          case TaskUpdateEffect.Noop =>
-            InstanceUpdateEffect.Noop(instance.instanceId)
-
-          case TaskUpdateEffect.Failure(cause) =>
-            InstanceUpdateEffect.Failure(cause)
-        }
-
-      case InstanceUpdateOperation.LaunchOnReservation(_, version, status, hostPorts) =>
+        tasks.find(_.taskId == taskId).map { task =>
+          val taskEffect = task.update(TaskUpdateOperation.MesosUpdate(status, mesosStatus, now))
+          taskEffect match {
+            case TaskUpdateEffect.Update(newTaskState) =>
+              val updated: Instance = updatedInstance(newTaskState, now)
+              val events = eventsGenerator.events(status, updated, Some(task), now)
+              updated.state.status match {
+                case t: Terminal =>
+                  InstanceUpdateEffect.Expunge(updated, events)
```

Comment:
We probably should expunge in the case of all tasks are terminal.
The instance can go to a terminal state in those cases:
```
    Error,
    Failed,
    Gone,
    Dropped
```
But not all tasks need to be in a terminal state.